File: core/src/main/java/org/kiwix/kiwixmobile/webserver/WebServerHelper.java
package org.kiwix.kiwixmobile.webserver;

import android.util.Log;
import androidx.annotation.NonNull;
import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import java.util.ArrayList;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import org.kiwix.kiwixlib.JNIKiwixException;
import org.kiwix.kiwixlib.JNIKiwixLibrary;
import org.kiwix.kiwixlib.JNIKiwixServer;
import org.kiwix.kiwixmobile.utils.ServerUtils;
import org.kiwix.kiwixmobile.wifi_hotspot.IpAddressCallbacks;

import static org.kiwix.kiwixmobile.utils.ServerUtils.INVALID_IP;

/**
 * WebServerHelper class is used to set up the suitable environment i.e. getting the
 * ip address and port no. before starting the WebServer
 * Created by Adeel Zafar on 18/07/2019.
 */

public class WebServerHelper {
  private static final String TAG = "WebServerHelper";
  private JNIKiwixLibrary kiwixLibrary;
  private JNIKiwixServer kiwixServer;
  private IpAddressCallbacks ipAddressCallbacks;
  private boolean isServerStarted;

  @Inject public WebServerHelper(@NonNull JNIKiwixLibrary kiwixLibrary,
    @NonNull JNIKiwixServer kiwixServer, @NonNull IpAddressCallbacks ipAddressCallbacks) {
    this.kiwixLibrary = kiwixLibrary;
    this.kiwixServer = kiwixServer;
    this.ipAddressCallbacks = ipAddressCallbacks;
  }

  public boolean startServerHelper(@NonNull ArrayList<String> selectedBooksPath) {
    String ip = ServerUtils.getIpAddress();
    if (ip.length() == 0) {
      return false;
    } else if (startAndroidWebServer(selectedBooksPath)) {
      return true;
    }
    return isServerStarted;
  }

  public void stopAndroidWebServer() {
    if (isServerStarted) {
      kiwixServer.stop();
      updateServerState(false);
    }
  }

  private boolean startAndroidWebServer(ArrayList<String> selectedBooksPath) {
    if (!isServerStarted) {
      int DEFAULT_PORT = 8080;
      ServerUtils.port = DEFAULT_PORT;
      for (String path : selectedBooksPath) {
        try {
          boolean isBookAdded = kiwixLibrary.addBook(path);
          Log.v(TAG, "isBookAdded: " + isBookAdded + path);
        } catch (JNIKiwixException e) {
          Log.v(TAG, "Couldn't add book " + path);
        }
      }
      kiwixServer.setPort(ServerUtils.port);
      updateServerState(kiwixServer.start());
      Log.v(TAG, "Server status" + isServerStarted);
    }
    return isServerStarted;
  }

  private void updateServerState(boolean isStarted) {
    isServerStarted = isStarted;
    ServerUtils.isServerStarted = isStarted;
  }

  //Keeps checking if hotspot has been turned using the ip address with an interval of 1 sec
  //If no ip is found after 15 seconds, dismisses the progress dialog
  public void pollForValidIpAddress() {
    Flowable.interval(1, TimeUnit.SECONDS)
      .map(__ -> ServerUtils.getIp())
      .filter(s -> s != INVALID_IP)
      .timeout(15, TimeUnit.SECONDS)
      .take(1)
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(
        s -> {
          ipAddressCallbacks.onIpAddressValid();
          Log.d(TAG, "onSuccess:  " + s);
        },
        e -> {
          Log.d(TAG, "Unable to turn on server", e);
          ipAddressCallbacks.onIpAddressInvalid();
        }
      );
  }
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/webserver/WebServerHelper.kt
/*
 * Kiwix Android
 * Copyright (c) 2023 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.webserver

import android.util.Log
import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import org.kiwix.kiwixmobile.core.R
import org.kiwix.kiwixmobile.core.utils.DEFAULT_PORT
import org.kiwix.kiwixmobile.core.utils.ServerUtils
import org.kiwix.kiwixmobile.core.utils.ServerUtils.INVALID_IP
import org.kiwix.kiwixmobile.core.utils.ServerUtils.getIp
import org.kiwix.kiwixmobile.core.utils.ServerUtils.getIpAddress
import org.kiwix.kiwixmobile.core.webserver.wifi_hotspot.IpAddressCallbacks
import org.kiwix.kiwixmobile.core.webserver.wifi_hotspot.ServerStatus
import java.util.concurrent.TimeUnit
import javax.inject.Inject

/**
 * WebServerHelper class is used to set up the suitable environment i.e. getting the
 * ip address and port no. before starting the WebServer
 * Created by Adeel Zafar on 18/07/2019.
 */
class WebServerHelper @Inject constructor(
  private val kiwixServerFactory: KiwixServer.Factory,
  private val ipAddressCallbacks: IpAddressCallbacks
) {
  private var kiwixServer: KiwixServer? = null
  private var isServerStarted = false
  private var validIpAddressDisposable: Disposable? = null

  fun startServerHelper(selectedBooksPath: ArrayList<String>): ServerStatus {
    val ip = getIpAddress()
    return if (ip.isNullOrEmpty()) {
      ServerStatus(false, R.string.error_ip_address_not_found)
    } else {
      startAndroidWebServer(selectedBooksPath)
    }
  }

  fun stopAndroidWebServer() {
    if (isServerStarted) {
      kiwixServer?.stopServer()
      updateServerState(false)
    }
  }

  private fun startAndroidWebServer(selectedBooksPath: ArrayList<String>): ServerStatus {
    var errorMessage: Int? = null
    if (!isServerStarted) {
      ServerUtils.port = DEFAULT_PORT
      kiwixServer = kiwixServerFactory.createKiwixServer(selectedBooksPath).also {
        updateServerState(it.startServer(ServerUtils.port))
        Log.d(TAG, "Server status$isServerStarted").also {
          if (!isServerStarted) {
            errorMessage = R.string.error_server_already_running
          }
        }
      }
    }
    return ServerStatus(isServerStarted, errorMessage)
  }

  private fun updateServerState(isStarted: Boolean) {
    isServerStarted = isStarted
    ServerUtils.isServerStarted = isStarted
  }

  // Keeps checking if hotspot has been turned using the ip address with an interval of 1 sec
  // If no ip is found after 15 seconds, dismisses the progress dialog
  @Suppress("MagicNumber")
  fun pollForValidIpAddress() {
    validIpAddressDisposable = Flowable.interval(1, TimeUnit.SECONDS)
      .map { getIp() }
      .filter { s: String? -> s != INVALID_IP }
      .timeout(15, TimeUnit.SECONDS)
      .take(1)
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(
        { s: String? ->
          ipAddressCallbacks.onIpAddressValid()
          Log.d(TAG, "onSuccess:  $s")
        }
      ) { e: Throwable? ->
        Log.d(TAG, "Unable to turn on server", e)
        ipAddressCallbacks.onIpAddressInvalid()
      }
  }

  fun dispose() {
    validIpAddressDisposable?.dispose()
  }

  companion object {
    private const val TAG = "WebServerHelper"
  }
}
